### PR TITLE
#39260 filter support for new shotgun model

### PR DESCRIPTION
**Pull request by manneohrstrom**
This updates the loader to work correctly with the new optimized shotgun model that defer loads things into memory.

- the filtering is 'capped' so that it doesn't kick in until you have typed 2 characters.
- the entire tree is loaded into memory at this point in order for the proxy model to operate.

### CODE REVIEW

File: python/tk_multi_loader/constants.py [hunk 1]
```diff
@@ -32,3 +32,7 @@
                           "version.Version.sg_status_list",
                           "created_by.HumanUser.image"
                           ]
+
+# left hand side tree view search only kicks in
+# after a certain number have been typed in.
+TREE_SEARCH_TRIGGER_LENGTH = 1
```
Comment (josh-t) [inline python/tk_multi_loader/constants.py]: Based on the PR summary, I would expect this to value to be `2`.

File: python/tk_multi_loader/dialog.py [hunk 2]
```diff
@@ -1276,7 +1278,9 @@ def _on_search_text_changed(self, pattern, tree_view, proxy_model):
         proxy_model.setFilterFixedString(pattern)
 
         # change UI decorations based on new pattern.
-        if pattern and len(pattern) > 0:
+        # for performance, make sure filtering only kicks in
+        # once we have typed at least one character
+        if pattern and len(pattern) > constants.TREE_SEARCH_TRIGGER_LENGTH:
```
Comment (josh-t) [inline python/tk_multi_loader/dialog.py]: Suggest making the value of `TREE_SEARCH_TRIGGER_LENGTH` be `2` and making this `>=` just for clarity. i.e. the tree search isn't triggered until the length is 2 or greater.

File: python/tk_multi_loader/dialog.py [hunk 3]
```diff
@@ -1276,7 +1278,9 @@ def _on_search_text_changed(self, pattern, tree_view, proxy_model):
         proxy_model.setFilterFixedString(pattern)
 
         # change UI decorations based on new pattern.
-        if pattern and len(pattern) > 0:
+        # for performance, make sure filtering only kicks in
+        # once we have typed at least one character
```
Comment (josh-t) [inline python/tk_multi_loader/dialog.py]: should be: typed at least **two** character**s**

File: python/tk_multi_loader/proxymodel_entity.py [hunk 4]
```diff
@@ -86,8 +88,29 @@ def setFilterFixedString(self, pattern):
         self._cache_hits = 0
         self._cache = {}
 
-        # call base class
-        return QtGui.QSortFilterProxyModel.setFilterFixedString(self, pattern)
+        if len(pattern) > constants.TREE_SEARCH_TRIGGER_LENGTH:
```
Comment (josh-t) [inline python/tk_multi_loader/proxymodel_entity.py]: Same comment here. `>=`.

File: python/tk_multi_loader/proxymodel_entity.py [hunk 5]
```diff
@@ -86,8 +88,29 @@ def setFilterFixedString(self, pattern):
...
+            # we have a search filter that is longer than one character.
+            # start filtering. Before we can filter, ensure that the entire
+            # data set is loaded in the tree.
+
+            def __fetch_more_r(index):
+                if self.canFetchMore(index):
+                    self.fetchMore(index)
```
Comment (josh-t) [inline python/tk_multi_loader/proxymodel_entity.py]: Same comment as other PR here. I'm assuming it's fine but this requires `fetchMore` to be synchronous in order for `rowCount` to be reliable right? Are we ok to make that assumption here?

File: python/tk_multi_loader/dialog.py (hunk 3 above)
Comment (manneohrstrom) [inline python/tk_multi_loader/dialog.py]: done

File: python/tk_multi_loader/dialog.py (hunk 2 above)
Comment (manneohrstrom) [inline python/tk_multi_loader/dialog.py]: done!

File: python/tk_multi_loader/proxymodel_entity.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tk_multi_loader/proxymodel_entity.py]: done!

File: python/tk_multi_loader/proxymodel_entity.py (hunk 5 above)
Comment (manneohrstrom) [inline python/tk_multi_loader/proxymodel_entity.py]: fixed!

**Comment by manneohrstrom** [comment]
ready for CR

**Pull request merged**